tests: mcuboot: pytest: Add image swap test with mcumgr

gchwier opened the pull request:
Added application based on SMP Server Sample. Application is build together with MCUboot using sysbuild and is flashed onto device in one step. Second application (smp_srv) is only build with sysbuild (not flashed). Tests are automated with pytest - new harness of Twister. Second application is uploaded by mcumgr into device and tested. Automated scenarios to test upgrades (image upload, test, revert, confirm), and to test downgrade prevention mechanism (separated in testcase.yaml, because needs to have changed some configuration options).
Added also scenario to test upgrade with an invalid signatures (for that case added helper method to wrap west sign command).

Run test with Twister:
`./zephyr/scripts/twister -vv --west-flash --enable-slow -T zephyr/tests/boot/with_mcumgr -p nrf9160dk_nrf9160 --device-testing --device-serial /dev/ttyACM0`
*mcumgr must be installed on the system ancd `mcumgr` command available.


Edit:
Second application is not built now. The image for upgrade is prepared using `west sign` command.

gopiotr commented on tests/boot/with_mcumgr/pytest/test_security.py:
```diff
@@ -0,0 +1,88 @@
+    """
+    assert wait_for_message(dut, "Launching primary slot application")
+
+    logger.info('Sign second image with wrong key')
```
This log is not proper for third scenario (with valid key). Maybe to more universal, simplify it to:
```diff
- logger.info('Sign second image with wrong key')
+ logger.info('Sign second image with key')
```

gopiotr commented on tests/boot/with_mcumgr/pytest/test_upgrade.py:
```diff
@@ -0,0 +1,125 @@
+
+import time
+import logging
+import pytest  # noqa # pylint: disable=unused-import
```
This import is unnecessary. Please remove it.

gopiotr commented on tests/boot/with_mcumgr/pytest/test_downgrade_prevention.py:
```diff
@@ -0,0 +1,52 @@
+
+import time
+import logging
+import pytest  # noqa # pylint: disable=unused-import
```
This import is unnecessary. Please remove it.

gopiotr commented on tests/boot/with_mcumgr/pytest/conftest.py:
```diff
@@ -0,0 +1,69 @@
+        command.extend(command_extra_args)
+
+    logger.info(f"CMD: {shlex.join(command)}")
+    output = check_output(command, text=True, timeout=10)
```
Please create dedicated variable for this timeout (like `west_sign_timeout = 10`), to avoid magic numbers in code.

gopiotr reviewed (commented):
In general LGTM, some minor comments below. Please consider also adding `README` file for those tests to highlight, that `mcumgr` is required, and with exemplary command how to run those tests (for me it wasn't so obvious that `sysbuild` requires using `--west-flash` option in Twister command - otherwise I get not so clear information that tests were skipped due to the `runtime filter`)

gopiotr commented:
One more observation - during running those tests by following command:
```
./scripts/twister -vv --inline-logs -p nrf52840dk_nrf52840 -T tests/boot/with_mcumgr --device-testing --device-serial /dev/ttyACM0 --west-flash
```
I got following warning from pytest at the end of test:
```
...
DEBUG   - PYTEST: =============================== warnings summary ===============================
DEBUG   - PYTEST: ../../.local/lib/python3.8/site-packages/_pytest/config/__init__.py:747
DEBUG   - PYTEST: /home/redbeard/.local/lib/python3.8/site-packages/_pytest/config/__init__.py:747: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: twister_harness.fixtures.mcumgr
DEBUG   - PYTEST: self.import_plugin(import_spec)
DEBUG   - PYTEST: -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
DEBUG   - PYTEST: - generated xml file: /home/redbeard/zephyrproject/zephyr/twister-out/nrf52840dk_nrf52840/tests/boot/with_mcumgr/boot.with_mcumgr.test_downgrade_prevention/report.xml -
DEBUG   - PYTEST: 1 passed, 5 deselected, 1 warning in 49.69s
...
```
It doesn't look so dangerous but it would be nice to know what is a root cause or what could be done to avoid it.

gchwier commented:
@gopiotr thanks for review!
Regarding `PytestAssertRewriteWarning`, I missed it, because I was testing with installed `pytest-twister-harness` plugin, then there is no warning, The rootcause is described here:
https://docs.pytest.org/en/latest/how-to/writing_plugins.html#requiring-loading-plugins-in-a-test-module-or-conftest-file
I will add `pytest.register_assert_rewrite('twister_harness.fixtures.mcumgr')` to `__init__.py` file and it should be enough for that warning.

gopiotr commented:
> @gopiotr thanks for review! Regarding `PytestAssertRewriteWarning`, I missed it, because I was testing with installed `pytest-twister-harness` plugin, then there is no warning, The rootcause is described here: https://docs.pytest.org/en/latest/how-to/writing_plugins.html#requiring-loading-plugins-in-a-test-module-or-conftest-file I will add `pytest.register_assert_rewrite('twister_harness.fixtures.mcumgr')` to `__init__.py` file and it should be enough for that warning.

Thanks for explanation - maybe do this adding `pytest.register_assert_rewrite('twister_harness.fixtures.mcumgr')` in separate commit?

gchwier commented:
> 
> Thanks for explanation - maybe do this adding `pytest.register_assert_rewrite('twister_harness.fixtures.mcumgr')` in separate commit?

I've added this to the current PR, but if we need this fix faster, I will move it later

gopiotr commented:
> I've added this to the current PR, but if we need this fix faster, I will move it later

I'm fine with have it in this PR - I propose to have it in separate commit - changes in `pytest-twister-harness` are connected with test tool, not directly with "mcuboot validate invalid signatures test", and thus now commit description for those changes are inadequate.

gchwier commented:
> 
> I'm fine with have it in this PR - I propose to have it in separate commit - changes in `pytest-twister-harness` are connected with test tool, not directly with "mcuboot validate invalid signatures test", and thus now commit description for those changes are inadequate.

Done. Moved to separate commit.

nordicjm commented on tests/boot/with_mcumgr/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
```
commit has typos:
*built
*smp_svr
*built
*revert

gchwier replied:
updated, thanks

PerMac commented:
@nordicjm @de-nordic Can you have a look if the test scenarios makes sense to you?

nordicjm commented on tests/boot/with_mcumgr/sysbuild.cmake:
```diff
@@ -0,0 +1,16 @@
+# Add the mcuboot key file to the secondary swapped app
+# This must be done here to ensure that the same key file is used for signing
+# both the primary and secondary apps
+set(smp_svr_CONFIG_MCUBOOT_SIGNATURE_KEY_FILE
+    \"${SB_CONFIG_BOOT_SIGNATURE_KEY_FILE}\" CACHE STRING
+    "Signature key file for signing" FORCE)
```
I don't see why this is needed, this should be done automatically by the sysbuild code if MCUboot is enabled

gchwier replied:
I was copied it from test `tests/boot/test_mcuboot`
https://github.com/zephyrproject-rtos/zephyr/blob/3e2765cc0da9f309b952ee31c9ff7cf6ad0a3b7a/tests/boot/test_mcuboot/sysbuild.cmake#L7

This is a secondary swapped app and when I removed this key file assignment, then I received:
`smp_svr/zephyr/.config:CONFIG_MCUBOOT_SIGNATURE_KEY_FILE=""`,
so it is not done automatically.

PerMac commented on tests/boot/with_mcumgr/testcase.yaml:
```diff
@@ -0,0 +1,30 @@
+common:
```
please add
`slow: true`
Since those tests require 600s to pass they fall into "slow" category which shouldn't be executed by twister by default.

nordicjm commented on tests/boot/with_mcumgr/overlay-mcuboot-version.conf:
```diff
@@ -0,0 +1,2 @@
+# Set version of MCUBOOT bypassing extra options to imgtool
+CONFIG_MCUBOOT_EXTRA_IMGTOOL_ARGS="--version 1.1.1+1"
```
Can do this with a `VERSION` file instead

gchwier replied:
I pass this by the `overlay-config`, because it is used only in one subtest in `testcase.yaml:boot.with_mcumgr.test_downgrade_prevention`:
https://github.com/zephyrproject-rtos/zephyr/blob/95e1df0d430ac90e3fa148dd67860a80df0bffa1/tests/boot/with_mcumgr/testcase.yaml#L30
I tried to pass it directly in `testcase.yaml` without overlay file, using `EXTRA_ARGS` ... but there was a problem with double quotes (they are removed by twister logic and not passed to cmake args).

gchwier replied:
I found there is also `MCUBOOT_IMGTOOL_SIGN_VERSION` config, but still I cannot use it in `testspec.yaml` as `extra_args`, because Twister removes double quotes (maybe issue to Twister to fix it),
Also I tried with `extra_config` option in Twister yaml, but looks, that this option does not support sysbuild yet (an overlay file is created and `SB_` and `mcuboot_` prefixes are not removed)

gchwier replied:
```
    extra_args:
      - mcuboot_CONFIG_MCUBOOT_DOWNGRADE_PREVENTION=y
    extra_configs:
      - CONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION="1.1.1+1"
```
but combination of extra_configs and extra_args works good. I removed this overlay file

nordicjm commented on tests/boot/with_mcumgr/pytest/mcumgr.py:
```diff
@@ -0,0 +1,66 @@
+        self.run_command('reset')
+
+    def image_upload(self, image: Path | str):
+        self.run_command(f'-t 60 image upload {image}')
```
60 seconds is a long time for a packet timeout, 25 would be more than sufficient

nordicjm commented on tests/boot/with_mcumgr/pytest/test_security.py:
```diff
@@ -0,0 +1,87 @@
+    'key_file',
+    [
+        '',
+        'root-ec-p256.pem',
```
pretty sure this won't work with mcuboot in zephyr at present due to ARM breaking imgtool, will submit a PR later today updating the mcuboot revision to include a fix

gchwier replied:
But this `key_file` is used only in `west sign` fixture:
https://github.com/zephyrproject-rtos/zephyr/blob/95e1df0d430ac90e3fa148dd67860a80df0bffa1/tests/boot/with_mcumgr/pytest/conftest.py#L30
and is not passed to west command if is empty.
West sign command is then called without `--key`, so the image will not be signed.

PerMac commented:
The comment is just for a minor detail. In general LGTM. I am not happy that it has to use 'platform_allow' but it seems to be the only way. There is already an open ticket for figuring out how to properly filter on mcuboot availability https://github.com/zephyrproject-rtos/zephyr/issues/58122

nordicjm commented on tests/boot/with_mcumgr/pytest/test_security.py:
```diff
@@ -0,0 +1,87 @@
+    valid_key = False
+    if key_file:
+        origin_key_file = find_in_config(
+            Path(dut.device_config.build_dir) / 'mcuboot' / 'zephyr' / '.config',
+            'CONFIG_BOOT_SIGNATURE_KEY_FILE'
```
this will not work, sysbuild will forcefully fix the wrong configuration with that from the sysbuild configuration

gchwier replied:
It works. Here I'm looking for the signature key file path in existing build output (not using any wrong `key_file` from pytest parameters).

fabiobaltieri commented:
@gchwier could you add a MAINTAINER entry for the new test directory so that followup PRs gets assigned by the bot? Or should this be under `tests/subsys/mgmt/` instead?

gchwier commented:
> @gchwier could you add a `MAINTAINERS.yml` entry for the new test directory so that followup PRs gets assigned by the bot? Or should this be under `tests/subsys/mgmt/` instead?

I do not know where is good place for these tests. Rather not `tests/subsys/mgmt/`, we are not testing `mcumgr`, but using it as a tool.
Here we want to test upgrade with MCUboot and we use sysbuild and pytest. In `MAINTAINERS.yml` I found section with `tests/boot/test_mcuboot/`, maybe could be better to extend that section.
https://github.com/zephyrproject-rtos/zephyr/blob/e1e4fcc70135e30e7634df80391edd77f5f3ba98/MAINTAINERS.yml#L3107

@de-nordic, @nordicjm, @d3zd3z what do you think about that?

nordicjm commented on tests/boot/with_mcumgr/sysbuild.cmake:
```diff
@@ -0,0 +1,16 @@
+  SOURCE_DIR ${ZEPHYR_BASE}/samples/subsys/mgmt/mcumgr/smp_svr
+)
+
+# Do not add the swapped app to the list of images to flash,
```
Now I see why it is not getting the signing key. This is not acceptable, it's hacking around sysbuild and calling things that sysbuild itself should be calling, understandable that you don't want to flash the image but need another way of doing it which sysbuild natively supports, not doing it by calling internal sysbuild cmake functions (I will take the task of looking into this)

gchwier replied:
I still don't understand. Let me explain:
- APP + MCUboot are built, signed and used to flash (like in [samples/application_development/sysbuild/with_mcuboot](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/application_development/sysbuild/with_mcuboot)
- `smp_svr` app is added as a third app, it is also built and signed, but **not added to the list of images to flash**. I used sysbuild here, just to have ready artefacts to use during test procedure: I need only `zephyr.signed.bin`, to send it with `mcumgr` and use to upgrade.
- `smp_svr` is added in similar way like the third app in  [tests/boot/test_mcuboot](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/boot/test_mcuboot), but there it is also added to the list of images to flash and is used when the image swap is requested. But this test is now skipped, because of #58080 

If using sysbuild to build a new app is a "hack" and is not acceptable, then I can remove third app (smp_svr) from sysbuild, then (in pytest scenario, after flashing) prepare image for upgrade using `west sign` (take `zephyr.signed.bin` from original APP and sign it with another version). But I think this will be also a bit tricky.

nordicjm replied:
The problem is by doing this, you are skipping lines 66 - 96 of https://github.com/zephyrproject-rtos/zephyr/blob/main/share/sysbuild/CMakeLists.txt#L66 when really, only line 95 should be skipped

PerMac replied:
@nordicjm IMO the question is what process we want to test and what must be done with sysbuild (maybe it was already answered?). Obviously, sysbuild must be used for building the bootloader and the application image. But in the process the same app image is used, the only difference is in the signing. Hence, is it mandatory to do duplicated builds with sysbuild just to have different "sign"?

gchwier replied:
@nordicjm I removed the `smp_svr`, now only APP + MCUboot are built with sysbuild. The image for upgrade is prepared now using `west sign` command, during pytest scenario.

de-nordic commented on tests/boot/with_mcumgr/pytest/test_downgrade_prevention.py:
```diff
@@ -0,0 +1,51 @@
+def test_downgrade_prevention(dut: Device, mcumgr: MCUmgr):
+    """
+    Verify that the application is not downgraded
+    1) Device flashed with MCUboot and an application that contains SMP server.
+       Version of MCUboot is 1.1.1
+    2) Prepare an update of an application containing the SMP server, where
+       MCUboot version is 0.0.0 (lower than version of the original app)
```
I do not think you mean "MCUboot version" but rather application version set for MCUboot or something like this. I doubt you are able to update MCUboot.

gopiotr commented on scripts/pylib/pytest-twister-harness/src/twister_harness/device/hardware_adapter.py:
```diff
@@ -170,6 +170,7 @@ def _disconnect_device(self) -> None:
             self._serial_connection.close()
             # self._serial_connection = None
             logger.debug('Closed serial connection for %s', serial_name)
+        self._serial_buffer.clear()
```
Please add/move `self._serial_buffer.clear()` to `_clear_internal_resources`

gchwier replied:
I will add clearing serial buffer to `_clear_internal_resources` method and remove it from `disconnect`. I found that after reconnecting, the `exec_command` method is looking for a prompt, that would be cleared from serial.
Thanks for finding it.

gopiotr commented on tests/boot/with_mcumgr/pytest/test_upgrade.py:
```diff
@@ -0,0 +1,209 @@
+
+
+@pytest.mark.parametrize(
+    'key_file', ['', 'root-ec-p256.pem'],
```
Please use `None` for `no_key` instead of empty string:
```
'key_file', [None, 'root-ec-p256.pem'],
```

nordicjm commented on tests/boot/with_mcumgr/pytest/mcumgr.py:
```diff
@@ -0,0 +1,66 @@
+
+if __name__ == '__main__':
+    logging.basicConfig(level=logging.DEBUG, format='%(levelname)-8s:  %(message)s')
+    mcumgr = McuMgr.create_for_serial('/dev/ttyACM0')
```
This is a problem, what might be ttyACM0 on one device isn't going to be ttyACM0 on another device, this is also linux-specific. What would be good is for it to use (assuming twister is used) the serial port that was provided to twister or the serial port from the twister hardware map file, not hardcoding to ttyACM0

gchwier replied:
This is outdated code and was added under `if __name__ == '__main__'` section (just added for some local testing). It is now removed, `mcumgr.py` is moved to `pytest-twister-harness` plugin and unit tests was added (it was changed in another PR #59480 and the current PR was rebased)

gchwier replied:
Squashed commits to not leave changes, that were later removed from that PR

nordicjm commented on scripts/pylib/pytest-twister-harness/tests/helpers/shell_mcuboot_command_parser_test.py:
```diff
@@ -0,0 +1,56 @@
+        confirmed: 0
+        primary area (1):
+        version: 0.0.2+0
+        image size: 68240
+        magic: good
+        swap type: test
+        copy done: set
+        image ok: unset
+        secondary area (3):
+        version: 0.0.0+0
+        image size: 68240
```
This is not going to work, it will work right now with the zephyr commit you used when you made this but can guarantee by zephyr 3.5 this will not work because the output size has changed slightly

gchwier replied:
It is only a unit test for `ShellMCUbootCommandParsed` helper class. When output of `mcuboot` command will change, it can be updated here, to test our helper class to be sure it still works as expected.
Some fields (like image size) are not checked here, there are added as an example of a real output.

PerMac replied:
@nordicjm Just to be clear. It is a mock of an output here, not an expected output from a test.

gchwier committed:
tests: mcuboot: pytest: Add image swap test with mcumgr

Added application based on SMP Server Sample. Application is built together
with MCUboot using sysbuild and is flashed onto device in one step.
Tests are automated with pytest - new harness of Twister.
The image for upgrade is prepared using west sign command
then is uploaded by mcumgr into device and tested.
Automated scenarios to test upgrade (image upload, test, revert, confirm),
to test downgrade prevention mechanism and to test upgrade with image,
that is signed with an invalid key.

Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>
